calyx-py/test/correctness/fifo.py

```diff
@@ -247,57 +290,93 @@ def add_main(prog, fifo):
 
-    ten_pushes = [
-        cb.invoke(
-            fifo,
-            in_pop=cb.const(1, 0),
-            in_push=cb.const(1, 1),
-            in_payload=cb.const(32, 100 + i),
-            ref_ans=ans,
-            ref_err=err,
-        )
-        for i in range(10)
+    # We will use the `invoke` method to call the `fifo` component.
+    fifo = main.cell("myfifo", fifo)
+    # The component takes as `ref` input:
+    err = main.mem_d1("err", 1, 1, 1)  # A flag to indicate an error
+    ans = main.mem_d1("ans", 32, 1, 32)  # A memory to hold the answer of a pop
+    # We will set up a while loop that runs over the command list, relaying
+    # the commands to the `fifo` component.
+    # It will run until the `err` flag is raised by the `fifo` component or
+    # until it reaches the end of the command list.
+
+    i = main.reg("i", 32)  # The index of the command we're currently processing
+    j = main.reg("j", 32)  # The index on the answer-list we'll write to
+    command = main.reg("command", 32)  # The command we're currently processing
+
+    zero_i = set_flag_reg(main, i, 0, "zero_i")  # zero out `i`
+    zero_j = set_flag_reg(main, j, 0, "zero_j")  # zero out `j`
+    incr_i = add_incr(main, i, "add3", "incr_i")  # i = i + 1
+    incr_j = add_incr(main, j, "add4", "incr_j")  # j = j + 1
+    err_eq_zero = add_eq_mem(main, err, 0, 0, "err_eq_0", 1)  # is the `err` flag down?
+    # i_eq_15 = add_eq(main, i.out, 15, "i_eq_15", 32)  # is `i` == 15?
+    # raise_err = mem_store(main, err, 0, cb.const(1, 1), "raise_err")  # set `err` to 1
+    read_command = mem_load_to_reg(main, commands, i.out, command, "read_command")
+    command_eq_zero = add_eq(main, command.out, 0, "command_eq_zero", 32)
+    write_ans = mem_store_from_mem(main, ans_mem, j.out, ans, 0, "write_ans")
+
+    main.control += [
+        zero_i,
+        zero_j,
+        cb.while_(
+            err_eq_zero[0].out,
+            err_eq_zero[1],  # Run while the `err` flag is down
+            [
+                read_command,  # Read the command at `i`
+                cb.if_(
+                    # Is this a pop or a push?
+                    command_eq_zero[0].out,
+                    command_eq_zero[1],
+                    [  # a pop
+                        cb.invoke(  # first we call pop
+                            fifo,
+                            in_pop=cb.const(1, 1),
+                            in_push=cb.const(1, 0),
+                            ref_ans=ans,
+                            ref_err=err,
+                        ),
+                        cb.if_(  # but was it successful?
+                            err_eq_zero[0].out,
+                            err_eq_zero[1],
+                            [  # yes, it was successful: write the answer to ans_mem
+                                write_ans,
+                                incr_j,
+                            ],
+                        ),
+                        # AM:
```

Comment:
It sounds like it _can_ be a register (see #1616) which would fix this issue I think. Regardless, I sorta don't get the issue. These are reads from the same memory, yes, but are not _parallel_ reads as far as I can tell. Either I'm missing something and Calyx is saving me, or Calyx is being overly cautious here...

Reply:
I suspect that you're running afoul of [the stability check](https://docs.calyxir.org/lang/ref.html#while) which assumes that the `comb group` is active throughout the body of the `if` or the `while`. It is a little annoying but we ended up figuring out that this restricts the language enough to reject some malformed programs (see https://github.com/cucapra/calyx/discussions/927).

Reply:
Gotcha, in this case I do need this behavior, so I'll just have to convert that `comb group` into a regular `group`. That is, instead of 
```
if err_eq_0.out with err_eq_0_group {...}
``` 

I'll change `err_eq_0_group` so it is not `comb` and then have 
```
[ 
  err_eq_0_group; if err_eq_0.out {...}
]
```
 in all instances. Does that sound kosher?

And actually now that I know the reason for the error, I'm guessing that moving to registers would not have helped, huh?

Reply:
Yeah, I think that should be fine! This is a general pattern that we use for conditions and is in fact exactly how `with` statements for `if` and `while` are compiled.

Reply:
Yep, confirming: the usual way we do things is that the `with` group is truly combinational; it should just do combinational operations like `std_lt` or whatever on stable values. Reading from a memory (by assigning to its address port) is not combinational. So the pattern is: read from a memory in a proper Calyx group activation, put that in a register, and then use a combinational group if necessary to compute stuff about that value you read.

Reply:
> Reading from a memory (by assigning to its address port) is not combinational.

This is true for `seq_mem`, which have a sequential interface for both reads and writes but not true for `std_mem` which have a combinational read interface and sequential write interface